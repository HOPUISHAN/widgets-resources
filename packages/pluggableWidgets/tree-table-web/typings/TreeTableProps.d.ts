/**
 * This file was generated from TreeTable.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, CSSProperties } from "react";
import { DynamicValue, ListValue, ListAttributeValue, ListWidgetValue, WebIcon } from "mendix";
import { Big } from "big.js";

export type ShowContentAsEnum = "attribute";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    showContentAs: ShowContentAsEnum;
    attribute?: ListAttributeValue<string | Big | boolean | Date>;
    header?: DynamicValue<string>;
    width: WidthEnum;
    size: number;
    alignment: AlignmentEnum;
}

export type ShowIconEnum = "left" | "right" | "no";

export interface ColumnsPreviewType {
    showContentAs: ShowContentAsEnum;
    attribute: string;
    header: string;
    width: WidthEnum;
    size: number | null;
    alignment: AlignmentEnum;
}

export interface TreeTableContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advancedMode: boolean;
    datasource: ListValue;
    parentIdAttribute?: ListAttributeValue<Big>;
    currentIdAttribute?: ListAttributeValue<Big>;
    hasChildren: boolean;
    startExpanded: boolean;
    children?: ListWidgetValue;
    animate: boolean;
    columns: ColumnsType[];
    showIcon: ShowIconEnum;
    expandedIcon?: DynamicValue<WebIcon>;
    collapsedIcon?: DynamicValue<WebIcon>;
    animateIcon: boolean;
}

export interface TreeTablePreviewProps {
    class: string;
    style: string;
    advancedMode: boolean;
    datasource: {} | { type: string } | null;
    parentIdAttribute: string;
    currentIdAttribute: string;
    hasChildren: boolean;
    startExpanded: boolean;
    children: { widgetCount: number; renderer: ComponentType<{ caption?: string }> };
    animate: boolean;
    columns: ColumnsPreviewType[];
    showIcon: ShowIconEnum;
    expandedIcon: { type: "glyph"; iconClass: string } | { type: "image"; imageUrl: string } | null;
    collapsedIcon: { type: "glyph"; iconClass: string } | { type: "image"; imageUrl: string } | null;
    animateIcon: boolean;
}
