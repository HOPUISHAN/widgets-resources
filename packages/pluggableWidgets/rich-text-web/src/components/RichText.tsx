import { createElement, useState, useRef, useEffect, ReactElement } from "react";
import { CKEditorType, CKEditorConfig, CKEditorEventAction, useCKEditor } from "ckeditor4-react";
import sanitizeHtml from "sanitize-html";
import classNames from "classnames";
import { ReadOnlyStyleEnum } from "../../typings/RichTextProps";

export interface RichTextSettings {
    config?: CKEditorConfig | null;
    type?: CKEditorType | null;
    sanitizeContent?: boolean;
}

export interface RichTextEditorProps {
    value: string;
    editorSettings: RichTextSettings;
    name?: string;
    onChange?: (value: string) => void;
    onKeyPress?: () => void;
    readOnlyStyle?: ReadOnlyStyleEnum;
}

export function RichTextEditor({
    value,
    name,
    editorSettings,
    onChange,
    onKeyPress,
    readOnlyStyle
}: RichTextEditorProps): ReactElement {
    const localEditorValueRef = useRef(value);
    // Use `useState` rather than `useRef` in order to trigger re-render.
    const [element, setElement] = useState<null | HTMLElement>(null);

    const { editor, status } = useCKEditor({
        ...editorSettings,
        element,
        editorUrl: `${window.mx.remoteUrl}widgets/ckeditor/ckeditor.js`,
        initContent: value,
        dispatchEvent: ({ type, payload }) => {
            if (type === CKEditorEventAction.key) {
                if (onKeyPress) {
                    onKeyPress();
                }
            }

            if (onChange && type === CKEditorEventAction.change) {
                const valueRaw = payload.editor.getData();
                const value = editorSettings.sanitizeContent ? sanitizeHtml(valueRaw) : valueRaw;

                localEditorValueRef.current = value;

                if (onChange) {
                    onChange(value);
                }
            }
        },
        subscribeTo: ["change", "key"]
    });

    useEffect(() => {
        if (editor) {
            if (localEditorValueRef.current !== value) {
                editor.setData(value);
                localEditorValueRef.current = value;
            }
        }
    }, [editor, localEditorValueRef, value]);

    return (
        <div
            className={classNames("widget-rich-text", readOnlyStyle && `editor-${readOnlyStyle}`)}
            style={{ width: editorSettings.config?.width, height: editorSettings.config?.height }}
        >
            <div id={name} ref={setElement} style={status !== "ready" ? { visibility: "hidden" } : undefined} />
        </div>
    );
}
