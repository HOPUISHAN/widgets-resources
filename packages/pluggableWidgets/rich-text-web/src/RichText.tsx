import { ReactNode, createElement, useMemo, useRef, useCallback } from "react";
import { RichTextEditor as RichTextComponent } from "./components/RichText";
import { RichTextContainerProps } from "../typings/RichTextProps";
import { debounce, executeAction } from "@mendix/piw-utils-internal";
import loadingCircleSvg from "./ui/loading-circle.svg";
import "./ui/RichText.scss";
import { useEditorSettings } from "./utils/useEditorSettings";

export default function RichText(props: RichTextContainerProps): ReactNode {
    const [editorSettings, editorKey] = useEditorSettings(props);
    const attributeRef = useRef(props.stringAttribute);
    attributeRef.current = props.stringAttribute;

    const onChange = useMemo(() => {
        return debounce((value: string) => {
            attributeRef.current.setValue(value);
            if (props.onChange) {
                executeAction(props.onChange);
            }
        }, 500);
    }, [props.onChange, attributeRef]);

    const onKeyPress = useCallback(() => executeAction(props.onKeyPress), [props.onKeyPress]);

    if (props.stringAttribute.status !== "available") {
        return <img src={loadingCircleSvg} className="widget-rich-text-loading-spinner" alt="" aria-hidden />;
    }

    return (
        <RichTextComponent
            key={editorKey}
            onChange={onChange}
            onKeyPress={onKeyPress}
            readOnlyStyle={props.stringAttribute.readOnly ? props.readOnlyStyle : undefined}
            editorSettings={editorSettings}
            value={props.stringAttribute.value ?? ""}
        />
    );
}
