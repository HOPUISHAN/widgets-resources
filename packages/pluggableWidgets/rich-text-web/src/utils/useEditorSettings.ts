import { useMemo } from "react";
import { CKEditorConfig } from "ckeditor4-react";
import { RichTextSettings } from "../components/RichText";
import { RichTextContainerProps } from "../../typings/RichTextProps";
import { getDimensions } from "@mendix/piw-utils-internal";
import { addPlugin, defineAdvancedGroups, getPreset } from "./ckeditorConfigs";
import { createCustomToolbar, GroupType, ToolbarItems } from "./ckeditorPresets";

type ConfigReducer = {
    (config: CKEditorConfig, props: RichTextContainerProps): CKEditorConfig;
    getDeps: (props: RichTextContainerProps) => any[];
};

const mapProps: ConfigReducer = (config, props) => {
    config.tabIndex = props.tabIndex;
    config.readOnly = props.stringAttribute.readOnly;
    config.disableNativeSpellChecker = !props.spellChecker;

    return config;
};

mapProps.getDeps = props => [props.tabIndex, props.stringAttribute.readOnly, props.spellChecker];

const editorWidth: ConfigReducer = (config, props) => {
    const { width, height } = getDimensions(props);
    config.width = width;
    config.height = height;

    return config;
};

editorWidth.getDeps = props => [props.width, props.widthUnit, props.height, props.heightUnit];

const contentFilter: ConfigReducer = (config, props) => {
    if (props.advancedContentFilter === "custom") {
        config.allowedContent = props.allowedContent;
        config.disallowedContent = props.disallowedContent;
    }

    return config;
};

contentFilter.getDeps = props => [props.advancedContentFilter];

const plugins: ConfigReducer = (config, props) => {
    if (props.codeHighlight) {
        addPlugin("codesnippet", config);
    }

    return config;
};

plugins.getDeps = props => [props.codeHighlight];

const toolbar: ConfigReducer = (config, props) => {
    const { advancedConfig, toolbarConfig, preset } = props;
    if (preset !== "custom") {
        return Object.assign(config, getPreset(preset));
    } else {
        const groupKeys = Object.keys(props).filter((key: string) => (key.includes("Group") ? key : null));
        let groupItems: ToolbarItems[] | string[];
        if (toolbarConfig === "basic") {
            groupItems = groupKeys
                .filter((groupName: GroupType) => props[groupName])
                .map((groupName: GroupType) =>
                    groupName.includes("separator") ? "/" : groupName.replace("Group", "").toLowerCase()
                );
        } else {
            groupItems = defineAdvancedGroups(advancedConfig);
        }

        return Object.assign(config, createCustomToolbar(groupItems, toolbarConfig === "basic"));
    }
};

toolbar.getDeps = props => [props.advancedConfig, props.toolbarConfig, props.preset];

const configReducers = [mapProps, editorWidth, toolbar, plugins, contentFilter];

function createConfig(props: RichTextContainerProps): CKEditorConfig {
    const initConfig: CKEditorConfig = {
        autoGrow_minHeight: 300,
        toolbarCanCollapse: true,
        autoGrow_onStartup: true
    };

    return configReducers.reduce((config, fn) => fn(config, props), initConfig);
}

type EditorSettingsKey = string;

type UseEditorSettingsResult = [RichTextSettings, EditorSettingsKey];

/**
 * This hook is intended to calculate CKEditor config.
 * As CKEditor don't allow dynamic configuration, along with
 * config we need calculate key. Key will be used to remount editor
 * every time config is changed.
 */
export function useEditorSettings(props: RichTextContainerProps): UseEditorSettingsResult {
    const getDependencies = (): any[] => {
        const selectors = [() => [props.editorType, props.sanitizeContent], ...configReducers.map(fn => fn.getDeps)];

        return selectors.map(fn => fn(props)).flat();
    };
    return useMemo(
        () => [
            {
                type: props.editorType,
                config: createConfig(props),
                sanitizeContent: props.sanitizeContent
            },
            Date.now().toString()
        ],
        // eslint-disable-next-line react-hooks/exhaustive-deps
        getDependencies()
    );
}
