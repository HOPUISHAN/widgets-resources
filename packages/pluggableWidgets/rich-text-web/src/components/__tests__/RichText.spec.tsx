import { createElement } from "react";
import "@testing-library/jest-dom";
import { RichTextEditor, RichTextEditorProps } from "../RichText";
import { getPreset, defineEnterMode, getToolbarGroupByName, defineAdvancedGroups } from "../../utils/ckeditorConfigs";
import { mount, ReactWrapper } from "enzyme";
import renderer from "react-test-renderer";
import { TOOLBAR_GROUP, ToolbarGroup } from "../../utils/ckeditorPresets";
import { AdvancedConfigType } from "../../../typings/RichTextProps";

declare global {
    interface Window {
        CKEDITOR: any;
    }
}

describe("RichText", () => {
    window.mx = {
        remoteUrl: ""
    };
    const defaultRichTextProps: RichTextEditorProps = {
        value: "Simple Text",
        name: "RichText_Div",
        readOnlyStyle: "text",
        editorSettings: {
            sanitizeContent: false,
            type: "classic",
            config: {
                advancedConfig: null,
                width: "480px",
                height: "320px",
                plugins: [],
                advancedContentFilter: null,
                readOnly: false,
                spellChecker: false,
                enterMode: "paragraph",
                shiftEnterMode: "paragraph",
                toolbar: getPreset("basic"),
                tabIndex: 1
            }
        }
    };

    function renderRichText(props = defaultRichTextProps): ReactWrapper {
        return mount(<RichTextEditor {...props} />);
    }

    it("render DOM structure", () => {
        const richText = renderer.create(<RichTextEditor {...defaultRichTextProps} />).toJSON();
        expect(richText).toMatchSnapshot();
    });

    it("renders dom elements with correct props", () => {
        const richText = renderRichText();
        expect(richText).toBeDefined();
        expect(richText.find(`#${defaultRichTextProps.name}`)).toBeDefined();

        const richTextElement = richText.find(".widget-rich-text");
        expect(richTextElement?.getElements().length).toBeGreaterThan(0);

        const styleProps = richTextElement.getElements()[0].props.style;
        expect(styleProps.width).toEqual(defaultRichTextProps.editorSettings.config!.width);
        expect(styleProps.height).toEqual(defaultRichTextProps.editorSettings.config!.height);

        expect(richText.find(`.editor-${defaultRichTextProps.readOnlyStyle}`)).toEqual({});
    });

    it("renders editor in read only mode with specified style", () => {
        const containerProps = { ...defaultRichTextProps, readOnly: true };
        const richText = renderRichText(containerProps);
        expect(richText.find(`.editor-${containerProps.readOnlyStyle}`).getElements().length).toBeGreaterThan(0);
    });
});

describe("CKEditor configuration", () => {
    it("returns correct value for enterMode", () => {
        let enterMode = defineEnterMode("paragraph");
        expect(enterMode).toEqual(1);

        enterMode = defineEnterMode("breakLines");
        expect(enterMode).toEqual(2);

        enterMode = defineEnterMode("blocks");
        expect(enterMode).toEqual(3);

        enterMode = defineEnterMode("");
        expect(enterMode).toEqual(1);
    });

    it("returns group named document", () => {
        const group = getToolbarGroupByName("document");
        expect(group).toEqual(TOOLBAR_GROUP[0]);
    });

    it("creates toolbar from advanced configuration", () => {
        const input: AdvancedConfigType[] = [
            { ctItemType: "About", ctItemToolbar: "toolbar1" },
            { ctItemType: "Anchor", ctItemToolbar: "toolbar2" }
        ];
        const grouped = defineAdvancedGroups(input);
        const result = [
            {
                name: "toolbar1",
                items: ["About"]
            },
            {
                name: "toolbar2",
                items: ["Anchor"]
            }
        ];
        expect(grouped).toMatchObject(result);
    });

    it("returns configuration object for basic preset", () => {
        const preset = "basic";
        const config = getPreset(preset);

        expect(config).toEqual({
            toolbarGroups: TOOLBAR_GROUP,
            removeButtons:
                "Source,Save,Templates,NewPage,ExportPdf,Preview,Print,Cut,Redo,Copy,Undo," +
                "Paste,PasteText,PasteFromWord,Find,Replace,SelectAll,Scayt," +
                "Form,Checkbox,Radio,TextField,Textarea,Select," +
                "Button,ImageButton,HiddenField,Strike,Underline,Subscript," +
                "Superscript,CopyFormatting,RemoveFormat,CreateDiv,Blockquote,JustifyLeft," +
                "BidiLtr,Image,Table,BidiRtl,JustifyCenter,JustifyRight,Language,JustifyBlock," +
                "HorizontalRule,SpecialChar,Smiley,PageBreak,Iframe,Styles,TextColor,BGColor," +
                "ShowBlocks,Maximize,Format,Font,FontSize,Anchor"
        });
    });

    it("returns configuration object for standard preset", () => {
        const preset = "standard";
        const config = getPreset(preset);

        const toolbarGroups: Array<ToolbarGroup | string> = [...TOOLBAR_GROUP];
        toolbarGroups.splice(4, 0, "/");
        expect(config).toEqual({
            toolbarGroups,
            removeButtons:
                "Save,Templates,NewPage,ExportPdf,Preview,Print,Find," +
                "Replace,SelectAll,Form,Checkbox,Radio,TextField,Textarea," +
                "Select,Button,ImageButton,HiddenField,Underline,Subscript," +
                "Superscript,CopyFormatting,CreateDiv,JustifyLeft,BidiLtr,BidiRtl," +
                "JustifyCenter,JustifyRight,Language,JustifyBlock,Smiley,PageBreak," +
                "Iframe,TextColor,BGColor,Font,FontSize,ShowBlocks"
        });
    });
});
