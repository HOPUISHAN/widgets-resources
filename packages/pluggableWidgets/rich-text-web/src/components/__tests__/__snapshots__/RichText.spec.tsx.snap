// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText render DOM structure 1`] = `
<div
  className="widget-rich-text editor-text"
  style={
    Object {
      "height": "320px",
      "width": "480px",
    }
  }
>
  <div
    id="RichText_Div"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  />
</div>
`;
